cluster-autoscaler/provisioningrequest/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package orchestrator
+
+import (
+	"errors"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/clusterstate"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/estimator"
+	"k8s.io/autoscaler/cluster-autoscaler/processors/status"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/apis/autoscaling.x-k8s.io/v1beta1"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/checkcapacity"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/conditions"
+	provreq_pods "k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/pods"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/provreqclient"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling"
+	ca_errors "k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/taints"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+
+	ca_processors "k8s.io/autoscaler/cluster-autoscaler/processors"
+	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+type scaleUpMode interface {
```

Comment:
ditto on the interface name.
Maybe `provisioningReqeustOrchestratorMode`

Reply:
I think orchestrator should be in general one, however we have pods and ProvReq, so we have two different orchestrators and a wrapper one (3 in total), so different ProvClasses I introduced as different scaleUp modes. I'm ok name it provisioningRequestScaleUpMode, but IMO it's too big and scaleUpMode is actually is not limited to provreq.

Reply:
I have a couple questions about the general delegation setup here:

1. Why is `WrapperOrchestrator` defined under `core` instead of `provisioningrequest`? It seems very ProvReq-specific.
2. Why do we want to have 2 levels of delegation here? Wouldn't it be easier to understand if we integrated delegating between ProvReq classes into `WrapperOrchestrator` (together with moving it under `provisioningrequest`)? The proposed setup is a ProvReq-specific Orchestrator delegating between the regular scale-up Orchestrator and another ProvReq-specific Orchestrator which then itself delegates between ProvReq-class-specific Orchestrators (which are called Modes for added confusion).
3. Is the only reason for introducing `scaleUpMode` not having to implement methods we don't care about (i.e. `ScaleUpToNodeGroupMinSize`)? If so, IMO `provReqOrchestrator` would be the best name. It's not a mode, it's a component performing something. Plus then the tie to the full `Orchestrator` interface is clear.
4. Have you thought about turning `WrapperOrchestrator` into a generic `DelegatingOrchestrator`? E.g. it would take a set of orchestrators and a set of functions that assign the pods to the orchestrators (e.g. by grouping by the `ProvisioningRequestPodAnnotationKey` value). And then we'd just have one Orchestrator per-ProvReq-class, and one level of delegation, handled under `core`. This would also make it easier to introduce new classes in downstream CA implementations.

Reply:
1) Agree, make sense to move wrapper to provisioningrequest package.
2) My goal was to make code easier to understand. I would like to have each object doing a small part of logic. I.e wrapper orchestrator is doing split between regular pods and ProvReq pods and choose the appropriate orchestrator. The provreq orchestrator may have a different implementation for other cloud providers (for example if someone implement their own ProvReq class). The current implementation of ProvReq orchestrator doesn't care about specific ProvReq classes. 
3) Yes and no :) Sure, provReq orchestrator doesn't implement `ScaleUpToNodeGroupMinSize` method. Another reason to introduce scaleUpMode is because in my mental model orchestrator is smth major and we shouldn't have many orchestrators. I'm fine to change the name to provReqOrchestrator but I think there will be confusion between interface name and provreqorchestrator struct. 
4) Mostly addressed in 2). I wouldn't mix main provReqOrchestator and wrapperOrchestrator. 
>This would also make it easier to introduce new classes in downstream CA implementations
I don't see how it will make easier to introduce new classes, it should be the same as it is now, no?

Reply:
Re 2.:

I get aiming for more objects with narrower responsibilities, but hopping between abstraction levels also has a readability cost. And the responsibilities for `wrapperOrchestrator` and `provReqOrchestator` actually seem quite similar to me - delegate pods to dependent orchestrators based on some pod properties. So while reading I'm actually wondering why they're separate, plus the hop has its own cost.

There are 2 meaningful differences I see between them:
* `wrapperOrchestrator` splits the pods and delegates to one of them only, while `provReqOrchestrator` leaves 
the splitting to the dependent orchestrators, and calls all of them.
Do we actually the splitting/delegating semantics to be different between the orchestrators? Wouldn't `provReqOrchestrator` work if it called only one of the dependent orchestrators (e.g. in an LRU fashion), and then `wrapperOrchestrator` could handle both cases?
* `provReqOrchestrator` books the capacity for all dependent orchestrators. Wouldn't it be better to block that capacity also from the regular scale-up? Then again, `wrapperOrchestrator` could handle both cases.

Re 3.:

Are there differences in responsibilities/expected semantics between `scaleUpMode.ScaleUp()` and `Orchestrator.ScaleUp()`? If there aren't, I'd want to highlight that `scaleUpMode` is a subset of `Orchestrator`, with the same expectations etc. - in which case I'd try to stay as close to the original `Orchestrator` name as possible. If there are meaningful differences, I'd change the interface/methods/arguments/comment to highlight these differences, or at least make it not seem like it's a subset of `Orchestrator`.

In any case I'd remove the "Mode" part of the name. It suggests that this component changes (parts of) the logic of some other component (and not that some other component will delegate the logic to this component). There's nothing changing in `provReqOrchestrator` logic if you provide it with different modes. The overall behavior might change, but not the `provReqOrchestrator` logic itself, so the "Mode" part seems misleading to me.

Re 4.:

Right now `provReqOrchestrator` doesn't allow downstream implementations to change the set of modes at all, and if it did it would also make sense to export this interface. You also said that other cloud providers might need to implement their own `provReqOrchestrator` if they want to implement their own class. This is basically what I'm trying to avoid - I'd like to have a delegating orchestrator generic and robust enough for cloud providers to use it to implement their own classes by just implementing a specific orchestrator for their pods, and providing a function to determine which pods are theirs. With what's proposed, they'd have to re-implement their own `provReqOrchestrator`, and then plumb it into `wrapperOrchestrator` carefully (splitting the provreq pods between different prov-req specific orchestrators), right?

We can make `provReqOrchestrator` configurable enough to achieve the same effect, but then we have to define clear semantics for this interface that the cloud-provider-specific implementations can rely on.

Reply:
Moved wrapper to provisioningrequest orchestrator package